src/analysis/rule_extraction.jl

```diff
@@ -148,11 +159,72 @@ function cluster_rules(clusters::Vector{T}, X::DataFrame, max_rules::T;
         error("Failed fitting SIRUS. Try increasing the number of scenarios/samples.")
     end
 
-    return rules(mach.fitresult)
+    if remove_duplicates
+        return _remove_duplicates(rules(mach.fitresult))
+    else
+        return rules(mach.fitresult)
+    end
 end
-function cluster_rules(clusters::Union{BitVector,Vector{Bool}}, X::DataFrame, max_rules::T;
-    kwargs...) where {T<:Int64}
-    return cluster_rules(convert.(Int64, clusters), X, max_rules; kwargs...)
+function cluster_rules(
+    clusters::Union{BitVector,Vector{Bool}}, X::DataFrame, max_rules::T;
+    seed::Int64=123, remove_duplicates::Bool=true, kwargs...
+)::Vector{Rule{Vector{Vector},Vector{Float64}}} where {T<:Int64}
+    return cluster_rules(
+        convert.(Int64, clusters), X, max_rules;
+        seed=seed, remove_duplicates=remove_duplicates, kwargs...
+    )
+end
+
+"""
+    _remove_duplicates(rules)::Vector{Rule{Vector{Vector},Vector{Float64}}}
+
+Returns `Vector{Rule}` without duplicate rules (if there's any). The criteria to choose
+which rule to keep is based on the rule consequence probability (the one with the highest
+probability is kept). If there are more than one rule with the same highest probability,
+then the first one is chosen.
+"""
+function _remove_duplicates(
+    rules::T
+)::T where {T<:Vector{Rule{Vector{Vector},Vector{Float64}}}}
+    # Extract subclauses from each rule without value
+    subclauses = join.([_strip_value.(r.condition) for r in rules], "_&_")
+    unique_subclauses = unique(subclauses)
+
+    # Return rules if there are no duplicate rules
+    n_unique_rules = length(unique_subclauses)
+    (n_unique_rules == length(rules)) && return rules
+
+    n_rules = length(rules)
+    n_duplicates = n_rules - n_unique_rules
+    @warn "$n_duplicates of $n_rules duplicated rules were found and are going to be removed."
```

Comment:
Wrap in `if` statement?

```julia
if n_duplicates > 0
```

Reply:
Line 199 guarantees that `n_duplicates > 0` when we get to this line.

Reply:
For some reason the line appears different now, probably due to some suggestion you made that I committed, but I was referring to this line: ` (n_unique_rules == length(rules)) && return rules`

Reply:
I see.

I would expand that into a full if statement for clarity - very easy for my eyes to gloss over it. I usually see those shortcut return styles used in the very first line where it's more obvious.

But this is a minor nitpick, feel free to ignore.